src/main/java/de/azapps/kafkabackup/restore/message/PartitionMessageWriterWorker.java

```diff
@@ -1,27 +1,127 @@
+  }
 
   @Override
   public void run() {
     try {
-      // TODO write messages to partition
-      // TODO store map of old and new offsets
-      // TODO prevent duplicates
+      log.info("Restoring messages for topic partition {}", topicPartitionToRestore);
       topicPartitionToRestore.setMessageRestorationStatus(MessageRestorationStatus.RUNNING);
-      Thread.sleep(20000L);
+
+      List<String> filesToRestore = restoreMessageS3Service
+          .getMessageBackupFileNames(topicPartitionToRestore.getTopicConfiguration().getTopicName(),
+              topicPartitionToRestore.getPartitionNumber());
+
+      log.debug("Got {} files to restore.", filesToRestore.size());
+
+      filesToRestore.forEach(this::restoreBackupFile);
+
       topicPartitionToRestore.setMessageRestorationStatus(MessageRestorationStatus.SUCCESS);
-    } catch (RuntimeException | InterruptedException exception) {
+
+      log.info("Message restoration succeeded for topic partition {}", topicPartitionToRestore.getTopicPartitionId());
+    } catch (RuntimeException ex) {
+      log.error(String.format("Message restoration for topic partition %s failed",
+          topicPartitionToRestore.getTopicPartitionId()),
+          ex);
       topicPartitionToRestore.setMessageRestorationStatus(MessageRestorationStatus.ERROR);
     }
   }
+
+  private void restoreBackupFile(String backupFileKey) {
+    List<Record> records = restoreMessageS3Service.readBatchFile(backupFileKey);
+    log.debug("Got {} messages from batch file {}.", records.size(), backupFileKey);
+
+    produceRecords(records);
+  }
+
+  private void initiateProducer(TopicPartitionToRestore topicPartitionToRestore,
+      RestoreArgsWrapper restoreArgsWrapper) {
+    Properties props = new Properties();
+    props.put("bootstrap.servers", restoreArgsWrapper.getKafkaBootstrapServers());
+    props.put("acks", "all");
+    props.put("retries", 1);
+    props.put("batch.size", PRODUCER_BATCH_SIZE);
+    props.put("linger.ms", 1);
+    props.put("buffer.memory", 33554432);
+    props.put("key.serializer", "org.apache.kafka.common.serialization.ByteArraySerializer");
+    props.put("value.serializer", "org.apache.kafka.common.serialization.ByteArraySerializer");
+
+    props.put("transactional.id", "restore-transactional-id" + topicPartitionToRestore.getTopicPartitionId());
+    this.kafkaProducer = new KafkaProducer<>(props);
+    kafkaProducer.initTransactions();
+  }
+
+  private void produceRecords(List<Record> recordsToProduce) {
+    try {
+      kafkaProducer.beginTransaction();
+      List<List<Record>> partitionedRecords = Lists.partition(recordsToProduce, PRODUCER_BATCH_SIZE);
+
+      for (List<Record> batch : partitionedRecords) {
+        List<Pair<Record, Future<RecordMetadata>>> futures = new ArrayList<>();
+        batch.forEach(record -> {
+          if (topicPartitionToRestore.getRestoredMessageInfoMap().containsKey(record.kafkaOffset())) {
```

Comment:
Sorry, if I read this code too sloppily (it's getting late 😑) but it seems we are populating this map in the future, when the record has been written. Doesn't that mean that until a record is written, we don't have it in the map, and therefore we don't detect the duplicate?

Reply:
Right, it worker when I was processing 1 by 1. Will fix it.